### PR TITLE
strip splitting for new ILD models based on dd4hep geometry

### PR DESCRIPTION
**Pull request by danieljeans**
BEGINRELEASENOTES
- add new Marlin processor DDStripSplitter, which implements the Strip Splitting Algorithm for eg a scintillator strip ECAL.
ENDRELEASENOTES

### CODE REVIEW

**Comment by petricm** [comment]
Please fix the warning in travis:
```
[66/113] Building CXX object CMakeFile...dEcalSplitter/src/DDStripSplitter.cc.o
In file included from ../Clustering/hybridEcalSplitter/src/DDStripSplitter.cc:1:0:
.././Clustering/hybridEcalSplitter/include/DDStripSplitter.h:29:7: warning: 'class DDStripSplitter' has pointer data members [-Weffc++]
 class DDStripSplitter : public Processor {
       ^~~~~~~~~~~~~~~
.././Clustering/hybridEcalSplitter/include/DDStripSplitter.h:29:7: warning:   but does not override 'DDStripSplitter(const DDStripSplitter&)' [-Weffc++]
.././Clustering/hybridEcalSplitter/include/DDStripSplitter.h:29:7: warning:   or 'operator=(const DDStripSplitter&)' [-Weffc++]
../Clustering/hybridEcalSplitter/src/DDStripSplitter.cc: In constructor 'DDStripSplitter::DDStripSplitter()':
../Clustering/hybridEcalSplitter/src/DDStripSplitter.cc:47:1: warning: 'DDStripSplitter::_cellIDLayerString' should be initialized in the member initialization list [-Weffc++]
 DDStripSplitter::DDStripSplitter() : Processor("DDStripSplitter") {
 ^~~~~~~~~~~~~~~
../Clustering/hybridEcalSplitter/src/DDStripSplitter.cc:47:1: warning: 'DDStripSplitter::_cellIDModuleString' should be initialized in the member initialization list [-Weffc++]
../Clustering/hybridEcalSplitter/src/DDStripSplitter.cc:47:1: warning: 'DDStripSplitter::_cellIDStaveString' should be initialized in the member initialization list [-Weffc++]
../Clustering/hybridEcalSplitter/src/DDStripSplitter.cc: In member function 'virtual void DDStripSplitter::processEvent(EVENT::LCEvent*)':
../Clustering/hybridEcalSplitter/src/DDStripSplitter.cc:210:11: warning: variable 'phi' set but not used [-Wunused-but-set-variable]
     float phi=-999;
           ^~~
../Clustering/hybridEcalSplitter/src/DDStripSplitter.cc:211:11: warning: variable 'theta' set but not used [-Wunused-but-set-variable]
     float theta=-999;
           ^~~~~
../Clustering/hybridEcalSplitter/src/DDStripSplitter.cc:233:31: warning: variable 'stripSplitter' set but not used [-Wunused-but-set-variable]
   std::vector <std::string> * stripSplitter;
                               ^~~~~~~~~~~~~
```

**Comment by danieljeans** [comment]
compiler warnings fixed.

File: Clustering/hybridEcalSplitter/readme.txt [hunk 1]
```diff
@@ -1,3 +1,25 @@
...
+[hybrid case ( mix of scint and si ) not yet implemented]
+
+Daniel Jeans, Nov 2018.
+
+example steering snippet:
+
+  <processor name="myDDStripSplitter" type="DDStripSplitter">
+    <parameter name="ECALcollections_tranStrips"> ECalBarrelScOddCollectionRec ECalEndcapScOddCollectionRec </parameter>
+    <parameter name="ECALcollections_longStrips"> ECalBarrelScEvenCollectionRec ECalEndcapScEvenCollectionRec </parameter>
+    <parameter name="cellSize0"> 5. </parameter>
```
Comment (rete) [inline Clustering/hybridEcalSplitter/readme.txt]: You should get this info from DD4hep segmentation I believe

File: Clustering/hybridEcalSplitter/src/DDStripSplitter.cc [hunk 2]
```diff
@@ -0,0 +1,728 @@
...
+
+
+void DDStripSplitter::processEvent( LCEvent * evt ) { 
+
+  if (_symmetry<0) setupGeometry();
+
+  if (_makePlots) {
+    // first fill some simple MC histos
+    try {
+      LCCollection * col = evt->getCollection("MCParticle");
```
Comment (rete) [inline Clustering/hybridEcalSplitter/src/DDStripSplitter.cc]: Can you put the collection name as a processor parameter too ?

File: Clustering/hybridEcalSplitter/src/DDStripSplitter.cc [hunk 3]
```diff
@@ -0,0 +1,728 @@
...
+
+  std::map < IMPL::LCCollectionVec*, std::string >::iterator ii;
+
+  // add the new collections to the event
+  for (ii=outputcolls.begin(); ii!=outputcolls.end(); ii++) {
+    evt->addCollection(ii->first, ii->second);
+  }
+
+  if (_saveIntersections) {
+    evt->addCollection(intersectionHits,"stripIntersections");
```
Comment (rete) [inline Clustering/hybridEcalSplitter/src/DDStripSplitter.cc]: Can you add processor parameters for these 3 collections ?

File: Clustering/hybridEcalSplitter/readme.txt (hunk 1 above)
Comment (danieljeans) [reply Clustering/hybridEcalSplitter/readme.txt]: @rete thanks for the detailed QA. done.

File: Clustering/hybridEcalSplitter/src/DDStripSplitter.cc (hunk 2 above)
Comment (danieljeans) [reply Clustering/hybridEcalSplitter/src/DDStripSplitter.cc]: done

File: Clustering/hybridEcalSplitter/src/DDStripSplitter.cc (hunk 3 above)
Comment (danieljeans) [reply Clustering/hybridEcalSplitter/src/DDStripSplitter.cc]: done

**Comment by danieljeans** [comment]
make more use of DD4hep geometry info to work out strip orientation in even/odd layers: user now passes hit collections in even / odd layers (previously user had to know which was "transverse" and "longitudinal").
Add a check that the geometry info we get from dd4hep is indeed strip-like.

these should make the processor more robust.

File: Clustering/hybridEcalSplitter/src/DDStripSplitter.cc [hunk 4]
```diff
@@ -0,0 +1,790 @@
...
+			  << "  found : " << theDetectors.size() << "  - first det: " << theDetectors.at(0).name() << std::endl ;
+
+  if( theDetectors.size()  != 1 ){
+    std::stringstream es ;
+    streamlog_out (ERROR) << " getExtension: selection is not unique (or empty)  includeFlag: " << dd4hep::DetType( includeFlag ) << " excludeFlag: " << dd4hep::DetType( excludeFlag )
+			  << " --- found detectors : " ;
+    for( unsigned i=0, N= theDetectors.size(); i<N ; ++i ){
+      streamlog_out (ERROR) << theDetectors.at(i).name() << ", " ;
+    }
+    assert(0);
```
Comment (rete) [inline Clustering/hybridEcalSplitter/src/DDStripSplitter.cc]: Can you replace all `assert()` calls by `throw marlin::StopProcessingException(this)` ?

File: Clustering/hybridEcalSplitter/src/DDStripSplitter.cc [hunk 5]
```diff
@@ -0,0 +1,790 @@
...
+  }
+  _symmetry = _caloGeomData->inner_symmetry;
+  _stripLength = _caloGeomData->layers[0].cellSize0;
+  _stripWidth  = _caloGeomData->layers[0].cellSize1;
+
+  // check that layer 1 is orthogonal to this one
+  float lay1_strSize0 = _caloGeomData->layers[1].cellSize0;
+  float lay1_strSize1 = _caloGeomData->layers[1].cellSize1;
+  if ( fabs(lay1_strSize0-_stripWidth)/_stripWidth > 0.05 || fabs(lay1_strSize1-_stripLength)/_stripLength > 0.05 ) {
+    cout << "doesn't look like a basic strip-based calo?? don't know how to deal with this geometry!" << endl;
```
Comment (rete) [inline Clustering/hybridEcalSplitter/src/DDStripSplitter.cc]: Can you replace all `cout` calls by `streamlog_out()` calls ?

File: Clustering/hybridEcalSplitter/src/DDStripSplitter.cc [hunk 6]
```diff
@@ -0,0 +1,790 @@
...
+	orientation = TRANSVERSE;
+      } else {
+	cout << "ERROR: strip orientation undefined!!" << endl;
+	assert(0);
+      }
+      toSplit = &_ecalCollectionsOddLayers;
+      break;
+    default:
+      streamlog_out ( ERROR ) << "ERROR crazy stuff!!! abandoning event..." << endl;
+      return;
```
Comment (rete) [inline Clustering/hybridEcalSplitter/src/DDStripSplitter.cc]: Here also you should throw an exception, either
- `throw marlin::SkipEventException(this)`
- `throw marlin::StopProcessingException(this)`

Comment (danieljeans) [reply Clustering/hybridEcalSplitter/src/DDStripSplitter.cc]: done (also added in a couple of other places)

File: Clustering/hybridEcalSplitter/src/DDStripSplitter.cc (hunk 5 above)
Comment (danieljeans) [reply Clustering/hybridEcalSplitter/src/DDStripSplitter.cc]: done

File: Clustering/hybridEcalSplitter/src/DDStripSplitter.cc (hunk 4 above)
Comment (danieljeans) [reply Clustering/hybridEcalSplitter/src/DDStripSplitter.cc]: done

**Pull request merged**